lib/Dialect/FIRRTL/Transforms/InjectDUTHierarchy.cpp

```diff
@@ -0,0 +1,230 @@
+
+    SymbolTable::setSymbolVisibility(newDUT, dut.getVisibility());
+    dut->setAttr("sym_name",
+                 b.getStringAttr(circuitNS.newName(wrapperName.getValue())));
+
+    // The original DUT module is now the wrapper.  The new module we just
+    // created becomse the DUT.
+    wrapper = dut;
+    dut = newDUT;
+
+    // Finish setting up the wrapper.  It can have no annotations.
+    AnnotationSet::removePortAnnotations(wrapper,
+                                         [](auto, auto) { return true; });
+    AnnotationSet::removeAnnotations(wrapper, [](auto) { return true; });
+  }
+
+  // Instantiate the wrapper inside the DUT and wire it up.
+  b.setInsertionPointToStart(dut.getBody());
+  ModuleNamespace dutNS(dut);
+  auto wrapperInst = b.create<InstanceOp>(
+      b.getUnknownLoc(), wrapper, wrapper.moduleName(), ArrayRef<Attribute>{},
+      ArrayRef<Attribute>{}, false,
+      b.getStringAttr(dutNS.newName(wrapper.moduleName())));
+  for (auto pair : llvm::enumerate(wrapperInst.getResults())) {
+    Value lhs = dut.getArgument(pair.index());
+    Value rhs = pair.value();
+    if (dut.getPortDirection(pair.index()) == Direction::In)
+      std::swap(lhs, rhs);
+    b.create<ConnectOp>(b.getUnknownLoc(), lhs, rhs);
+  }
+
+  // Get ready to update non-local annotations (NLAs).  This requires both an
+  // NLA table and knowledge of what the DUT's port symbols are.
+  auto &nlaTable = getAnalysis<NLATable>();
+  DenseSet<Attribute> dutPortSyms;
+  for (auto port : dut.getPorts()) {
+    if (!port.sym)
+      continue;
+    dutPortSyms.insert(port.sym);
+  }
+
+  // Update NLAs invovle the DUT.  There are three cases to consider:
+  //   1. The DUT or a DUT port is a leaf ref.  Do nothing.
+  //   2. The DUT is the root.  Update the root module to be the wrapper.
+  //   3. The NLA passes through the DUT.  Remove the original InnerRef and
+  //      replace it with two InnerRefs: (1) on the DUT and (2) one the wrapper.
+  for (auto nla : llvm::make_early_inc_range(nlaTable.lookup(dut))) {
+    // The leaf ref is the DUT or a DUT port.
+    if (nla.leafMod() == dut.moduleName())
+      if (nla.isModule() || dutPortSyms.contains(nla.ref()))
+        continue;
+
+    // The DUT is the root module.
+    auto namepath = nla.namepath().getValue();
+    if (nla.root() == dut.getNameAttr()) {
+      auto tail = namepath.drop_front();
+      SmallVector<Attribute> newNamepath({hw::InnerRefAttr::get(
+          wrapper.getNameAttr(),
+          namepath.front().cast<hw::InnerRefAttr>().getName())});
+      newNamepath.append(tail.begin(), tail.end());
+      nla->setAttr("namepath", b.getArrayAttr(newNamepath));
+      continue;
+    }
+
+    // The NLA passes through the DUT.
+    size_t nlaIdx = 0;
+    for (auto attr : namepath) {
+      auto innerRef = attr.cast<hw::InnerRefAttr>();
+      if (innerRef.getModule() == dut.moduleName())
+        break;
+      nlaIdx++;
+    }
```

Comment:
I think change `dut.moduleName()` to `dut.moduleNameAttr()`, otherwise the attribute turns into a `StringRef`.

might be able to make something like this work:
```c++
auto nlaIdx = std::distance(llvm::find_if(namepath, [&](Attribute attr) {
  return attr.cast<hw::InnerRefAttr>().getModule == dut.moduleNameAttr());
  }));
```

Reply:
Nice suggestion. I was struggling with a clean way to do this without the `size_t nlaIdx = 0` mutation. 👍